### ISSUE TITLE
Is it better to enable SCSI WCE as default on Linux ?

### ISSUE THREAD

**Issue by homerl** [openzfs/zfs#11134]
### Describe the feature would like to see added to OpenZFS
Because the SCSI WCE default was disabled.

### How will this feature improve OpenZFS?
Is it better to add a udev policy? 
```bash
$ cat 50-zfs-udev.rules
# Reset some HDD vendor default value
ACTION=="add", KERNEL=="sd*[!0-9]", RUN+="/bin/sh -c 'smartctl --set=wcache,on /dev/%k '"
```

**Comment by amotin** [comment]
ZFS sends cache flushes everywhere it really needs. It does not benefit from disabling the write caches.

**Comment by richardelling** [comment]
WCE should be a site policy and ZFS shouldn't change site policies.

**Comment by homerl** [comment]
Hi @richardelling @amotin 
     Thank you.
     Sorry about my English.

 I mean the WCE disable in Linux by default.
```
sd 1:0:171:0: [sdfg] Write cache: disabled, read cache: enabled, supports DPO and FUA
sd 1:0:173:0: [sdfi] Write cache: disabled, read cache: enabled, supports DPO and FUA
```

After Add the 50-zfs-udev.rules, enable the WCE by default in Linux, The WCE could increase the write performance.
```
sd 1:0:44:0: [sdas] Write cache: enabled, read cache: enabled, supports DPO and FUA
sd 1:0:49:0: [sdax] Write cache: enabled, read cache: enabled, supports DPO and FUA
```

**Comment by shodanshok** [comment]
@homerl I fail to see how this is ZFS related. Rather, it seems a block device policy which should be set/unset regardless of the filesystem used. Am I missing something?

**Comment by homerl** [comment]
@shodanshok Yes, it depends on the SCSI device firmware default setting. 
If OpenZFS could set it by default that means to improve the default write performance for Linux users : )

**Comment by ZNikke** [comment]
Write cache is a changeable drive setting. WCE is commonly enabled when buying a HDD vendor branded drive, and disabled when buying a server vendor branded drive as they expect write caching to be handled by a hardware RAID controller anyway. Notably most modern RAID controllers allows users to control the drive write cache setting.

If you replace the raid controller role with zfs, I understand the opinion that zfs should provide the drive write cache disable/enable setting. OTOH, I understand that people might not want/expect zfs to change this, but the "site policy" argument is a bit flawed since if zfs allowed you to set it then you'd truly have a site policy.

As a side note, I seem to remember that ZFS on Solaris actually enabled the write cache on drives that had it disabled but supported DPO/FUA.

The middle ground is probably to document this, and provide options (sdparm howto, udev helper script, etc) that makes this easy to change where it's safe.

I for one have a bunch of SAS HDDs with WCE disabled that supports DPO/FUA. Using sdparm to enable WCE and save it permanently is one option, but then if someone replaces a broken HDD in the future the replacement might have WCE disabled which will cause a performance degradation... Going the udev route is probably better in that regard, but then I'll have to ensure that it doesn't fiddle with non-ZFS devices... Just being able to say "hey zfs, drivewritecache=enable please" would make this much easier.

**Comment by ZNikke** [comment]
For those in a similar situation as us, this is what we currently use to enable the write cache on devices supporting DPO/FUA:

```
# HDDs sold by server vendors usually have the drive write cache disabled,
# as they're intended to sit behind a RAID controller.
#
# ZFS is designed to properly manage drive caches and flushes write
# operations where needed, so it's safe to enable it.
#
# We restrain ourselves to only touch devices which gets assigned a vdev
# name by our vdev_id.conf.
#
SUBSYSTEM=="block", ACTION=="add|change", ENV{ID_VDEV}=="?*", ENV{DEVTYPE}=="disk", RUN+="/bin/sh -c \"sdparm --page=ca --long $devnode | grep -q DPOFUA=1 && sdparm --get=WCE -H $devnode | grep -q '^0x00 0x01' && sdparm --quiet --set=WCE $devnode || true\""
```

While the udev rule could make the changes permanent, we choose not to do that since we believe that automated rules should not change device-permanent settings.

If anyone knows a better way to figure out DPOFUA support that would be appreciated, this relies on what seems to be undocumented reporting by sdparm.

And yes, we have devices that has write cache but does not support DPOFUA, mostly older gen SSD:s.

**Comment by richardelling** [comment]
Actually, Sun used to ship HDDs with write cache disabled because UFS is susceptible to power loss events.
Hence, when ZFS sees that it has the whole disk, it tries to enable the write cache on illumos/Solaris systems.
Since those days (turn of the century) most if not all vendors ship with write cache enabled and there are 
many cases where the write cache cannot be disabled.

As for losing data, the cache that is important is RAM. Forcing write cache to be disabled on disks doesn't 
solve the problem. So it also means that FUA doesn't improve the reliability in any significant way. These 
things just make HDDs slower. For SSDs, it is not uncommon for cache flushes and FUA to be no-ops in 
enterprise-grade devices.

Also, the last time I checked, Linux disables FUA by default on SATA devices anyway, because of flaky reliability.

Therefore, I don't recommend changing these settings for the general case.

**Comment by ZNikke** [comment]
For clarity, our HDDs are various Dell branded SAS HDDs that used to sit behind a Dell PERC H730P RAID controller (ie Dell branded LSI/Avago/nameoftheday megaraid). These come with the write cache disabled, and I've seen this with other server vendor HDDs as well.

Same models directly from HDD vendors do ship with the write cache enabled.

**Comment by richardelling** [comment]
I believe you are making the assumption that since the drives behind the megaraid had write cache disabled, then the drives as shipped also had write cache disabled and the megaraid did not change that setting. It makes sense for megaraid to do this, due to it not being a transactional object store like ZFS.

**Comment by ZNikke** [comment]
Yes, it makes sense for megaraid to not change the write cache setting, as it's the controller default (at least in these particular machines), typically looks like:

```
# /opt/MegaRAID/perccli/perccli64 /c0/v0 show all | grep Disk.Cache
Disk Cache Policy = Disk's Default
```

**Comment by homerl** [comment]
> Actually, Sun used to ship HDDs with write cache disabled because UFS is susceptible to power loss events. Hence, when ZFS sees that it has the whole disk, it tries to enable the write cache on illumos/Solaris systems. Since those days (turn of the century) most if not all vendors ship with write cache enabled and there are many cases where the write cache cannot be disabled.
> 
> As for losing data, the cache that is important is RAM. Forcing write cache to be disabled on disks doesn't solve the problem. So it also means that FUA doesn't improve the reliability in any significant way. These things just make HDDs slower. For SSDs, it is not uncommon for cache flushes and FUA to be no-ops in enterprise-grade devices.
> 
> Also, the last time I checked, Linux disables FUA by default on SATA devices anyway, because of flaky reliability.
> 
> Therefore, I don't recommend changing these settings for the general case.

The WCE could increase the random write IOPS, Enable the WCE if the bottleneck in the random write, and if not, you could disable it.   
And WCE is only in SCSI/SAS, the same with SATA write cache.   
```
/**
 * blk_queue_flush - configure queue's cache flush capability
 * @q:          the request queue for the device
 * @flush:      0, REQ_FLUSH or REQ_FLUSH | REQ_FUA
 *
 * Tell block layer cache flush capability of @q.  If it supports
 * flushing, REQ_FLUSH should be set.  If it supports bypassing
 * write cache for individual writes, REQ_FUA should be set.
 */
void blk_queue_flush(struct request_queue *q, unsigned int flush)
{
        WARN_ON_ONCE(flush & ~(REQ_FLUSH | REQ_FUA));

        if (WARN_ON_ONCE(!(flush & REQ_FLUSH) && (flush & REQ_FUA)))
                flush &= ~REQ_FUA;

        q->flush_flags = flush & (REQ_FLUSH | REQ_FUA);
}
```

**Comment by akostadinov** [comment]
I'm not sure it's very relevant to the discussion, but looking at latest 20TB "enterprise" SATA HDDs (e.g. WD gold, Toshiba MG10 series), they seem to write the cache to platters by using the power of the spinning down platters. So decision about write cache is also about hardware.
